### ISSUE TITLE
Deal with list substitution

### ISSUE THREAD

**Issue by casey** [casey/just#208]
Currently, a `+` argument is a string consisting of the original arguments separated by spaces.

This presents difficulties, since the shell will re-split the text. If the argument contains intra-argument spaces, bad things can happen. 

For example, with the following justfile:

```make
compile flags+:
  cc {{flags}}
```
Invoking just with:

`just compile -DVAR='hello, world'`

Will run the command:

`cc -DVAR=hello, world`

Since the single quotes were evaluated in the calling shell, the command line will be parsed by the subshell invoked by just as `cc`, `-DVAR=hello,`, and `world`, which is clearly not what was intended.

So, given this problem, here's what I'm thinking of for a solution:

I'll add functions, with the first one being a `quote()` function. Quote may be applied to any variable, like so: `quote(variable)`. In a recipe, one would use it like this:

```make
compile flags+:
  cc {{quote(flags)}}
```

It will surround each element of a list with quotes, escaping existing quotes and backslashes as necessary.

I haven't decided on whether to use quotes or double quotes. I think it mostly depends on which has the most consistent behavior between `dash` and `bash`.

This complicates the type system a little bit, since we now must distinguish between lists and strings. I propose that all variables are now lists of strings, and so existing variables are now single-element lists. So, the quote function can be applied to all variables.

**Comment by casey** [comment]
@ssokolow, we discussed this a bit in #114

What do you think?

**Comment by ssokolow** [comment]
My main concern is that it feels like it might be a case of getting the wrong result from all of the right parts... possibly for unavoidable reasons.

1. The need to opt into quoting *is* in line with bourne-style shell semantics.
2. Functions like `quote()` *are* in line with Django/Jinja/Twig/etc.-style templating syntax. (though, typically as filters like `{{flags|quote}}`)
3. There's a limit to how far you want to drift from Bourne semantics for familiarity's sake.
4. You do want to retain backwards compatibility with existing Justfiles.

Despite all that, I have this nagging feeling that it's Bourne shell's biggest footgun, writ more verbose.  

(In my dash and bash scripts, I end up writing constructs like `cd "$(dirname "$(readlink -f "$0")")"` because splitting should be opt in, rather than opt out. That's actually one of the biggest reasons I mostly use Python for my command scripting these days and never write Bourne without Syntastic and shellcheck installed.)

If it weren't for the backwards compatibility issue, I'd say to make variables like auto-quoted and use the function to opt into unquoted behaviour.

**Comment by casey** [comment]
I agree that it's an ugly solution, but with the Bourne shell's terrible splitting semantics, it feels like it's the best worst option.

Auto-quoting variables is interesting, but requiring users to opt-out of quoting feels weird. I guess it would be an `unquote()` function?  `quote()` feels clearer, since you take a list of strings as input and you get a list of strings out, but with the quotes added. `unquote()` would be more of a type system mechanism thing, where you take a list of strings-that-are-destined-to-be-quoted in and get a list of strings-that-are-not-destined-to-be-quoted out, which seems harder to explain and understand.

Alternatively, we can build a time machine, go back to 1989 and pick up Tom Duff – right after he wrote the rc shell, which has much better splitting and quoting behavior – and then head to 1977 so he can have a chat with Stephen Bourne and we can avoid this whole mess.

**Comment by ssokolow** [comment]
I'd think of it more in terms of `split()` (ie. all variables are lists which get serialized to strings using quotes but most have length `1`) with `split()` splitting up each list member using Bourne quoting/escaping semantics.

Then, you'd get this intuitive-for-anything-but-bash behaviour:
`my_command {{split(args)}} {{infile_containing_spaces}} {{outfile_containing_spaces}}`

(You'd want to split using something like the `shlex` crate since some commands have options which take a quoted `argv` string that they'll parse or `shell()` themselves.)

**Comment by ssokolow** [comment]
...and I'd argue that, since `{{variable_name}}` already isn't Bourne syntax, it'd be reasonable to give it different quoting behaviour. It's not as if you have to redefine how quoting works for `$variable_name`.

Heck, I actually found it confusing and UN-intuitive when something which was obviously a pre-processing step separate from the Bourne shell evaluation with syntax inspired by an HTML templating engine which auto-escapes didn't auto-escape in this context.

**Comment by valscion** [comment]
Hi there :relaxed: I'm wondering is there any way to currently work around this lack of functionality in `just`?

We have more than 50 just tasks currently for various unrelated tasks, so we have split them up to multiple directories to simulate "namespaces". We've got a `Justfile` in the root of our application that looks something like this:

```
help:
  @just -l

admin +args='help':
  @just --working-directory `pwd` --justfile tasks/admin/Justfile {{ args }}

aws +args='help':
  @just --working-directory `pwd` --justfile tasks/aws/Justfile {{ args }}

chores +args='help':
  @just --working-directory `pwd` --justfile tasks/chores/Justfile {{ args }}
```

So we end up going back to `just` with the given variadic arguments.

This all works nicely up until we have some script that wants arguments to be passed like `--something="foo bar"`. We've yet to find a nice way to keep the spaces in the `--something` in place up until the final execution, as the quotes for `--something` are "lost" at the first step of this process.

**Comment by valscion** [comment]
Thanks for getting back to me so soon 😊. `just` is already pretty amazing and it works splendidly for 98% of our use cases.

Issues with quoting are definitely one of the biggest annoyances with shells and difficult to solve. I didn't expect you to have any solution but hey, it was worth a try 😃

Thanks for this tool, it has been really helpful in automating our mundane admin and chore tasks ☺️

**Comment by valscion** [comment]
Ok, now I think I'm in a happy place. I can write commands like this:

```
just admin fix-venue-location staging --where-venue-like="'Food Camp / %'" --lat=60.1817 --lon=24.522
```

We'll merely have to use both single and double quotes for arguments containing spaces. That's completely reasonable, as there's (as far as I understand) no way for `just` to understand the importance of these quotes.

I only need to use this kind of syntax in various places when I'm going deeper in to other tasks. Yes, it's additional glue code and I wish I didn't need to write that, but it will suffice.

```make
# in root Justfile
admin +args='help':
  #!/usr/bin/env bash
  set -eo pipefail
  args_as_is=$(cat <<EOF;
  {{args}}
  EOF)
  just --working-directory `pwd` --justfile tasks/admin/Justfile $args_as_is

# in tasks/admin/Justfile
fix-venue-location env='staging' +ARGS='--usage':
  #!/usr/bin/env bash
  set -eo pipefail
  args_as_is=$(cat <<EOF;
  {{ARGS}}
  EOF)
  just admin _run-ruby-script-using-confirmation {{env}} fix_venue_location $args_as_is

_run-ruby-script-using-confirmation env script_name +ARGS:
  #!/usr/bin/env bash
  set -eo pipefail
  args_as_is=$(cat <<EOF;
  {{ARGS}}
  EOF)
  # ...and so on
```

**Comment by casey** [comment]
It's definitely not ideal, but I'm glad that you were able to find a workaround.

I just opened #383, which I hope eventually leads to just natively supporting submodules and subcommands, which I think would be really cool

**Comment by kwshi** [comment]
My two cents:

1. I like the `quote()` approach the most (speaking of which, it can/should also be applicable to plain string arguments).
   - I don't think quoting-by-default/requiring opt-out of quoting is a particularly great idea, because I think that behavior is too "magical"/implicit and can cause even more unexpected issues depending on where it is used (e.g., what if they were trying to get the `$*` behavior and already trying to substitute it into quotes?  Then automatic quoting would unexpectedly introduce additional quotes canceling the existing ones, fudging up the syntax).  
   
     I believe that keeping the default behavior the explicit, straightforward, and dumb is in line with the philosophy of avoiding magic-induced, Make-esque idiosyncrasies.

2. In the absence of this feature, there is a partial workaround, I think: `set positional-arguments` would allow a single list to be passed to the recipe w/o escaping issues (since they are passed directly as `$1`, `$2`, etc.).

**Comment by casey** [comment]
@kwshi I agree with your thoughts on `quote()`. One thing that's stopped me from adding `quote()` is would ideally be a function from `List<String> → List<String>`, and Just currently only has values of type string. (Variadic arguments are just joined with spaces into a single string when being evaluated and inserted into a recipe's scope.)

Ideally, Just would have a static type system, so `quote()` could distinguish between items of type `String` and `List<String>`. But, a static type system is quite a large undertaking, and I've been too lazy to actually do it.

Also, @valscion, there's a new feature which you might find useful for passing arguments that would otherwise need to be quoted. (@kwshi, thanks for reminding me!)

If you add `set positional-arguments` to a justfile, recipe arguments will be made available as positional arguments.

For example:

```
set positional-arguments

variadic +arg:
  echo $@ # values of `arg`

single arg:
  echo $1 # value of `arg`
```

The shell won't interpret quotes, spaces, or other special characters inside of positional argument values, and will pass them to sub-commands unmmolested.

Another thing you can do is add a `$` in front of an argument, which will cause it to be made available as an environment variable, with much the same effect:

```
single $arg:
  echo $arg # value of `arg`
```

I think that this might help with your use cases.

**Comment by valscion** [comment]
Cool, thanks for letting me know of these new features! I'll keep them in mind the next time I'm figuring out our Justfile usages ☺️

**Comment by bckohan** [comment]
I was able to get this working on any systems with a basic bourne shell. The strategy is to use positional arguments and wrap them in extra quotes. In my case I needed to pass to a python script so I do it like this: 

```just
set script-interpreter := ['uv', 'run', '--script']
set positional-arguments := true

[script]
_manage *COMMAND:
    import os
    import sys
    import shlex
    from pathlib import Path
    from django.core import management
    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "tests.settings.base")
    management.execute_from_command_line(["just manage", *shlex.split("{{ COMMAND }}")])

# run the django admin
[macos]
[linux]
manage *COMMAND:
    #! /bin/sh
    ARGS_LIST=""
    for arg in "$@"; do
    case "$arg" in
        *[[:space:]]*)
            # Argument contains a space; wrap it in quotes.
            ARGS_LIST="${ARGS_LIST} \"\'${arg}\'\""
            ;;
        *)
            # No spaces, print as-is.
            ARGS_LIST="${ARGS_LIST} ${arg}"
            ;;
        esac
    done
    eval "just _manage $ARGS_LIST"
```

This allows things like this to work:

``` just manage shellcompletion install --manage-script="just manage"```

**I have been unable to get this to work on windows - positional-arguments does not seem to work for passing arguments correctly in powershell :(**

Anyway - just adding my 2p that this would be great to have! I get that its a big task.

Thankyou!

**Comment by luisrayas3** [comment]
Hi, thanks for writing `just`, I'm a fan.
That said I believe this is `just`'s greatest deficiency. I believe quoting-by-default is correct (the current behavior was very surprising to me) and indeed all variables should just be lists (modern shells like `fish` do exactly this and everything gets much better). Different quoting/escaping styles can always be handled by a `set`. I understand that for backward compatibility opt-in is the way so `quote` would be fine. When `just2` drops maybe it could become the default 😄.

Anyway more than provide my opinion on the matter I wanted to clarify and bump only real general-purpose workaround which is somewhat buried in this thread: double-quoting into `just`:

```sh
just test -m "'x  y  z'"
```

It's important that the double-quote go outside in case the argument contains substitutions (`"'$x &(y) z'"` works as you'd expect)

Hope this helps others.

**Comment by asibahi** [comment]
As I mentioned in #3147 Perhaps a solution for this would be either to special case `—-` or add a new flag for just called `—-raw` or `++` or something, with the idea that anything after that would be quoted before being sent to the underlying shell (provided of course there is an `args+` or something like it in th recipe).

Maybe controlled by a setting if this is deemed to be too breaking a change.

**Comment by LordGaav** [comment]
I'm using multiple recipes that wrap a call to another tool and insert some environment variables and other parameters. I was confused why Just seems to be eating the quotes around a call like this:

```
$ just alembic fooapi -- revision --autogenerate -m "Migrate foo to bar"
# The quotes around the parameter to `-m` are gone inside the shell recipe.
```

For me, using the `--` as a marker to interpret everything after that verbatim, and not mangle any quotes.

The entire recipe is like this, I was trying to mimic the behaviour of a bash script and how we used `$@` there with `{{ args }}` .

```
[arg("configfile", long="configfile")]
[doc("Run alembic on a single project")]
alembic project configfile="settings.yaml" *args:
    #!/bin/sh
    set -euo pipefail
    projectdir=$({{ uv }} pip list --editable --format json | {{ jq }} -r '.[] | select(.name == "{{ project }}") | .editable_project_location')
    configpath="${projectdir}/{{ configfile }}"
    if [ -f "${configpath}" ]; then
        echo "{{ YELLOW }}Found omniconf configuration, setting OMNICONF_YAML_FILENAME env variable to: '${configpath}'{{ NORMAL }}"
        export OMNICONF_YAML_FILENAME="${configpath}"
    else
        echo "{{ RED }}No omniconf configuration found for file '{{ configfile }}'{{ NORMAL }}"
    fi
    echo "{{ YELLOW }}Running alembic for {{ project }}{{ NORMAL }}"
    set -x
    exec {{ uv }} run -p python3 --no-managed-python --frozen --directory "$projectdir" --package "{{ project }}" alembic {{ args }}
```

**Comment by casey** [comment]
I just merged #3372, which adds an unstable `set lists` setting and changes the behavior of `quote()`. This follows an earlier change which made all values lists of strings, #3362, although this did not result in any observable changes.

Without `set lists`, variadic arguments to recipes are joined into a single space-separated string, and there are no changes in behavior. With `set lists`, variadic arguments are kept as lists, and additionally, when a list is passed to `quote()`, each list element is quoted individually.

This means that given the following justfile:

```
set lists

foo *args:
  echo "{{ quote(args) }}"
```

You'll get the following output:

- `foo bar`: `'bar'`
- `foo bar baz`: `'bar' 'baz'`
- `foo`: ``

Notice that when passed an empty list, `quote()` returns the empty string, not `''`.

The reason that `set lists` is unstable is because there are a huge number of places where we'd have list-aware behavior. For example `append('.txt', args)` would ideally append `.txt` to each argument, too many to land in a single PR or release.

**Comment by casey** [comment]
I created #3377 to track `set lists` stabilization.
